### ISSUE TITLE
MINOR: First shortcut after restarting service doesn't remap

### ISSUE THREAD

**Issue by RedBearAK** [rbreaves/kinto#511]
This is technically minor but annoying, and I'd like to see if one of us can fix it someday, if that's possible. Just ran into it several times today, so I'm going to document it. 

I probably only really notice this enough to get annoyed by it because I restart Kinto so frequently to test new configs. It's sort of like the issue where the first character you type is "eaten" at the login screen. (Actually it might be really closely related to that phenomenon, now that I think about it.) 

Basically, after restarting from the tray icon menu, the first time I press something like Alt+Tab, I actually get a Tab character (if I'm in a text editor of some kind). As if the modifier that was pressed first was ignored (or not intercepted). You could call it the "first character fails" phenomenon. 

Same thing happens in the Kinto GUI, but the Tab character gets inserted in the GUI's terminal view, since the GUI window has the focus. Whereas the tray menu doesn't steal focus from the current window, so the stray Tab gets inserted in the application I'm using. 

This also, of course, affects Kinto every single time it first starts, not just restarting. The very first attempt to do anything like Alt+Tab away from the GUI window (or Cmd+Q to close it) always fails. I didn't really pay much attention to it for a long time, because after that first key press it starts working, but I'm pretty sure this has always been the case, even before the GUI existed. It's very consistent in the way that it always fails and then starts working. 

My working theory on the cause is that `xkeysnail` starts up with no knowledge about the current window's WM_CLASS, and needs some kind of prompt to acquire the WM_CLASS class of the focused window before it starts working, and that the "prompt" is the first keycode being intercepted. 

If this theory is correct, fixing this could be as simple as... maybe artificially passing a "harmless" shortcut (or even just a single character) that doesn't do anything to `xkeysnail` immediately after restarting the service, before the user has a chance to attempt any shortcut. 

I can confirm that doing something as simple as hitting the Shift key after restarting the service completely bypasses this issue. So if I can form a habit of always remembering to hit Shift before reflexively trying Alt+Tab or Cmd+Q, theoretically I'll never see the issue again. 

I wonder if this could help cure the issue of character eating at the login screen, even though detecting the logout is still difficult, and dependent on the DE. What if every now and then a background process sent `xkeysnail` a keycode that doesn't do anything (and therefore couldn't really interfere with normal keyboard usage), but that ends up triggering `xkeysnail` to check the current WM_CLASS. This "heartbeat" character would cause the service to transparently go into the failed state at the login screen because the WM_CLASS can't be acquired, and the character eating phenomenon would stop before the user had a chance to encounter it while entering the password. 

If this worked, you wouldn't even need to try and detect the logout event anymore. It would work with all DEs and login managers. 

Just had a fuzzy thought about maybe using the "empty WM_CLASS" fix to partially bypass this behavior. It probably won't work. But I'll look into it.

**Comment by RedBearAK** [comment]
@rbreaves 

Update: The cause and the cure are probably not what I thought. It is only modifier keys that get "lost" or "eaten" upon first starting or restarting the `xkeysnail` service. Non-modifier keys like letters seem to pass through just fine after a Kinto restart, and any key press whatsoever (modifier or not) will completely bypass the issue. It's just that if the first key press after the service starts is a modifier key, it will disappear, as if you never pressed the key. And if it was part of a shortcut combo the associated keyboard shortcut will just insert the remainder of the shortcut into the current window. 

So when I try to do Alt+Q, I always get a "Q" in the window, and Alt+Tab types a Tab character. 

But, if you send a single modifier key press (or any key press, technically) immediately after starting the service, that should still cure the behavior. Even though I kind of had it all wrong. Sending a modifier key should guarantee that it won't get inserted into the current window, because it will be "lost" at the same time that it triggers `xkeysnail` to start intercepting modifier key presses properly.

**Comment by RedBearAK** [comment]
No worries. Just an annoyance that I'd like users not to be exposed to, eventually. It's a "speed bump" that I realized I was hitting on every single Kinto install. 

I may even look into the `xkeysnail` code a bit mayself and see if I can see some ridiculously obvious way to fix this in whatever function initiates the modifier remap. You never know. 

Never noticed it was just modifier keys because 99% of the time I attempt to test a shortcut combo immediately after a restart. 

See you next time.

**Comment by RedBearAK** [comment]
@joshgoebel 

>Did you ever figure this out or make any headway? 

I did not. I still can't make heads or tails of most of the `xkeysnail` code. Still don't have enough context to understand what most of it is doing. Sorry. 

>Do you see this behavior on my branch or 0.4.0?

I'm not certain that I did or didn't. I'll have to get back on that machine and test again to see specifically if the first Alt+Tab works correctly. But I would think you'd have noticed if you had the problem. Usually the terminal would "bonk" at me the first time I tried to switch away to another app. It's fairly obvious. But a GUI text editor app will insert a Tab character instead of Alt+Tabbing. A little more subtle.

**Comment by joshgoebel** [comment]
> But I would think you'd have noticed if you had the problem.

Yeah, I'd think so... only curious cause it's such a weird bug. :)  Are the events in the log?

**Comment by joshgoebel** [comment]
> I still can't make heads or tails of most of the xkeysnail code. 

If you want a 5 or 10 minute tour sometime, let me know.

**Comment by RedBearAK** [comment]
@joshgoebel 

Looks like there is no apparent issue with the first modifier press getting lost. With your branch, that is. Don't know about mainline `xkeysnail`. 

I can Alt+Tab or Cmd+T or such immediately after starting the process, without any issue or error "bonks" in the terminal. 

Here is output from a short run of the Kinto branch where I tried to Alt+Tab away from the terminal (it failed), then Alt+Tabbed away and back (successfully), and killed the process with Cmd+dot (Ctrl+C). 

You can see the caret, bracket, and tab space inserted where it failed. I don't see an immediate difference in the printed output for the first Alt+Tab as compared to the other Alt+Tab events. The first Alt press just seems to get ignored. 

```
Watching keyboard devices plug in
Printing key/action:     Key.RIGHT_CTRL             Action.PRESS
WM_CLASS 'Gnome-terminal' | active keymaps = [Magic shortcut to kill Kinto, Special overrides for terminals, General GUI, Wordwise - not vscode, terminals]
RC-TAB
Printing key/action:     Key.RIGHT_CTRL             Action.RELEASE
Printing key/action:     Key.LEFT_ALT             Action.PRESS
Printing key/action:     Key.TAB             Action.PRESS
^[	Printing key/action:     Key.TAB             Action.RELEASE
Printing key/action:     Key.RIGHT_CTRL             Action.PRESS
Printing key/action:     Key.RIGHT_CTRL             Action.RELEASE
Printing key/action:     Key.LEFT_ALT             Action.RELEASE
Printing key/action:     Key.RIGHT_CTRL             Action.RELEASE
Printing key/action:     Key.RIGHT_CTRL             Action.PRESS
WM_CLASS 'Gnome-terminal' | active keymaps = [Magic shortcut to kill Kinto, Special overrides for terminals, General GUI, Wordwise - not vscode, terminals]
RC-TAB
Printing key/action:     Key.RIGHT_CTRL             Action.RELEASE
Printing key/action:     Key.LEFT_ALT             Action.PRESS
Printing key/action:     Key.TAB             Action.PRESS
Printing key/action:     Key.TAB             Action.RELEASE
Printing key/action:     Key.RIGHT_CTRL             Action.PRESS
Printing key/action:     Key.RIGHT_CTRL             Action.RELEASE
Printing key/action:     Key.LEFT_ALT             Action.RELEASE
Printing key/action:     Key.RIGHT_CTRL             Action.RELEASE
Printing key/action:     Key.RIGHT_CTRL             Action.PRESS
WM_CLASS 'firefox' | active keymaps = [Anonymous keymap, General Web Browsers, Anonymous keymap, General GUI, Wordwise - not vscode]
RC-TAB
Printing key/action:     Key.RIGHT_CTRL             Action.RELEASE
Printing key/action:     Key.LEFT_ALT             Action.PRESS
Printing key/action:     Key.TAB             Action.PRESS
Printing key/action:     Key.TAB             Action.RELEASE
Printing key/action:     Key.RIGHT_CTRL             Action.PRESS
Printing key/action:     Key.RIGHT_CTRL             Action.RELEASE
Printing key/action:     Key.LEFT_ALT             Action.RELEASE
Printing key/action:     Key.RIGHT_CTRL             Action.RELEASE
Printing key/action:     Key.RIGHT_CTRL             Action.PRESS
WM_CLASS 'Gnome-terminal' | active keymaps = [Magic shortcut to kill Kinto, Special overrides for terminals, General GUI, Wordwise - not vscode, terminals]
RC-DOT
Printing key/action:     Key.RIGHT_CTRL             Action.RELEASE
Printing key/action:     Key.LEFT_CTRL             Action.PRESS
Printing key/action:     Key.C             Action.PRESS
^CReceived an interrupt, exiting.
```
